pkg/operator/endpoint_status_builder.go

```diff
@@ -68,24 +68,155 @@ func (b *scrapeEndpointBuilder) add(target *prometheusv1.TargetsResult) error {
 	return nil
 }
 
+type ScrapeJobKind uint8
+
+const (
+	ScrapeJobInvalid ScrapeJobKind = iota
+	ScrapeJobKubelet
+	ScrapeJobPodMonitoring
+	ScrapeJobClusterPodMonitoring
+	ScrapeJobNodeMonitoring
+)
+
+func probeKindToString(kind ScrapeJobKind) string {
+	switch kind {
+	case ScrapeJobKubelet:
+		return "kubelet"
+	case ScrapeJobPodMonitoring:
+		return "PodMonitoring"
+	case ScrapeJobClusterPodMonitoring:
+		return "ClusterPodMonitoring"
+	case ScrapeJobNodeMonitoring:
+		return "NodeMonitoring"
+	default:
+		return ""
+	}
+}
+
+func parseScrapeKind(kind string) (ScrapeJobKind, error) {
+	switch kind {
+	case probeKindToString(ScrapeJobKubelet):
+		return ScrapeJobKubelet, nil
+	case probeKindToString(ScrapeJobPodMonitoring):
+		return ScrapeJobPodMonitoring, nil
+	case probeKindToString(ScrapeJobClusterPodMonitoring):
+		return ScrapeJobClusterPodMonitoring, nil
+	case probeKindToString(ScrapeJobNodeMonitoring):
+		return ScrapeJobNodeMonitoring, nil
+	default:
+		return ScrapeJobInvalid, fmt.Errorf("invalid scrape kind %q", kind)
+	}
+}
+
+func parseScrapeJob(pool string) (monitoringv1.PodMonitoringCRD, error) {
+	split := strings.Split(pool, "/")
+	kind, err := parseScrapeKind(split[0])
+	if err != nil {
+		return nil, err
+	}
+	switch kind {
+	case ScrapeJobKubelet:
+		if len(split) != 1 {
+			return nil, fmt.Errorf("invalid Kubelet scrape job format %q", pool)
+		}
+		return nil, nil
+	case ScrapeJobPodMonitoring:
+		if len(split) != 3 {
+			return nil, fmt.Errorf("invalid PodMonitoring scrape job format %q", pool)
+		}
+		return &monitoringv1.PodMonitoring{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: split[1],
+				Name:      split[2],
+			},
+		}, nil
+	case ScrapeJobClusterPodMonitoring:
+		if len(split) != 2 {
+			return nil, fmt.Errorf("invalid ClusterPodMonitoring scrape job format %q", pool)
+		}
+		return &monitoringv1.ClusterPodMonitoring{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: split[1],
+			},
+		}, nil
+	case ScrapeJobNodeMonitoring:
+		if len(split) != 2 {
+			return nil, fmt.Errorf("invalid NodeMonitoring scrape job format %q", pool)
+		}
+		return nil, nil
+	default:
+		return nil, fmt.Errorf("unknown scrape kind %q", split[0])
+	}
+}
+
+type scrapePool struct {
+	job   string
+	group string
+}
+
+func parseScrapePool(pool string) (scrapePool, error) {
```

Comment:
what does `scrapePool` look like vs `parseScrapeJob`?

Reply:
It's confusing but let me explain the terminology.

If you look at Prometheus' configuration, a scrape configuration has a job name. That job name is the fully qualified name, e.g. for PodMonitoring it's `PodMonitoring/namespace/name/endpoint`. When Prometheus actually does the scraping, this is called the "scrape pool". That's because the job actually could scrape multiple targets, so it actually scrapes a pool of resources. All of the targets have the same "job" label however, which is the PodMonitoring name. I use the same terminology of scrape pool and job here. The scrape pool contains the job name.

We have to split up this information twice. Remember for PodMonitoring, we want to patch target status for all endpoints into the same PodMonitoring status. Thus, for `PodMonitoring/namespace/name/endpoint`, `scrapePool.job=PodMonitoring/namespace/name` and `scrapePool.group=endpoint`.

After we grouped all endpoints together, we have to know which PodMonitoring to save the target status too. That's where we split up the job `PodMonitoring/namespace/name` and turn that into a PodMonitoring object.

Let me know if that's unclear.

Reply:
Updated the names to make it clearer!

Reply:
Great explanation - thanks!

Could you add a comment in-code to clarify? Especially with the example - makes it really clear.

Reply:
Added the example in-code. Thanks!

Reply:
Thanks for adding the comment!